flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchCompactCoordinator.java

```diff
@@ -0,0 +1,155 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.file.table.batch.compact;
+
+import org.apache.flink.connector.file.table.BinPacking;
+import org.apache.flink.connector.file.table.stream.compact.CompactMessages.CompactionUnit;
+import org.apache.flink.connector.file.table.stream.compact.CompactMessages.CoordinatorInput;
+import org.apache.flink.connector.file.table.stream.compact.CompactMessages.CoordinatorOutput;
+import org.apache.flink.connector.file.table.stream.compact.CompactMessages.InputFile;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.function.SupplierWithException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Coordinator for compaction in batch mode. It will collect the written files in {@link
+ * BatchFileWriter} and determine whether to compact files or not as well as what files should be
+ * merged into a single file.
+ */
+public class BatchCompactCoordinator extends AbstractStreamOperator<CoordinatorOutput>
+        implements OneInputStreamOperator<CoordinatorInput, CoordinatorOutput>, BoundedOneInput {
+
+    private static final long serialVersionUID = 1L;
+
+    private final SupplierWithException<FileSystem, IOException> fsFactory;
+    private final long compactAverageSize;
+    private final long compactTargetSize;
+
+    private transient FileSystem fs;
+    // the mapping from written partitions to the corresponding files.
+    private transient Map<String, List<Path>> inputFiles;
+
+    public BatchCompactCoordinator(
+            SupplierWithException<FileSystem, IOException> fsFactory,
+            long compactAverageSize,
+            long compactTargetSize) {
+        this.fsFactory = fsFactory;
+        this.compactAverageSize = compactAverageSize;
+        this.compactTargetSize = compactTargetSize;
+    }
+
+    @Override
+    public void initializeState(StateInitializationContext context) throws Exception {
+        super.initializeState(context);
```

Comment:
will this `initializeState` be invoked by streaming job? BTW, I think this should be put in `open()`?

Reply:
dito.